app/services/authentication.rb

```diff
@@ -0,0 +1,45 @@
+class Authentication
+  def self.call(encoded_token)
+    new(encoded_token).call
+  end
+
+  def initialize(encoded_token)
+    @encoded_token = encoded_token
+  end
+
+  def call
+    user_by_sign_in_user_id || user_by_email
+  end
+
+private
+
+  def decoded_token
+    (decoded_token, _algorithm) = JWT.decode(
+      @encoded_token,
+      Settings.authentication.secret,
+      Settings.authentication.algorithm
+    )
```

Comment:
I don't know how expensive JWT.decode is, but would it be worth memoizing `decoded_token` here? It just seems to me that this could potentially run a few times in the span of a request.

Reply:
It's not possible to use restructuring and memoizing at the same time. So I opted for this, as it keeps the readability, which would otherwise suffer. What do you think? Got any better ideas?

Reply:
How about doing it in `initialize`?

Reply:
I'm not sure how that solves the issue? Also shouldn't do "work" in initialisers.

Reply:
That could be solved in `initialize` by setting `@decoded_token`, but I kind of agree with not doing work in the initialiser. So, alternatively:

```
   def decoded_token
     @decoded_token ||= JWT.decode(
      encoded_token,
      Settings.authentication.secret,
      Settings.authentication.algorithm
    ).first
  end
```

The destructuring is a nice to have, but ultimately disposable I think.

Reply:
Yeah, I opted for the destructuring because it's more readable and it should be quite rare to have to decode this many times, and it seems like a relatively cheap operation so I can't see it being an issue. But I don't have strong feelings either way